### PR TITLE
[1.8.x]Add support for union type request body mapping to Ballerina to OpenAPI spec generation

### PR DESCRIPTION
**Pull request by lnash94**
Fix https://github.com/wso2-enterprise/internal-support-ballerina/issues/443
Implement the union-type mapping for request body in OpenAPI spec generation

## Automation tests
 - Unit tests - added
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: openapi-bal-service/src/main/java/io/ballerina/openapi/converter/service/OpenAPIRequestBodyMapper.java [hunk 1]
```diff
@@ -152,62 +160,176 @@ public void handlePayloadAnnotation(RequiredParameterNode payloadNode, Map<Strin
...
      *</pre>
      */
-    private void handleSinglePayloadType(RequiredParameterNode payloadNode, Map<String, Schema> schema,
-                                         RequestBody bodyParameter, String customMediaPrefix) {
+    private void handlePayloadType(TypeDescriptorNode payloadNode, Map<String, Schema> schema,
+                                   RequestBody bodyParameter) {
+        SyntaxKind kind = payloadNode.kind();
+        String mediaTypeString = getMediaTypeForSyntaxKind(payloadNode);
+        if (mediaTypeString != null || payloadNode.kind() == SyntaxKind.UNION_TYPE_DESC ||
+                payloadNode.kind() == SyntaxKind.QUALIFIED_NAME_REFERENCE ||
```
Comment (dilanSachi) [inline openapi-bal-service/src/main/java/io/ballerina/openapi/converter/service/OpenAPIRequestBodyMapper.java:168]:
```suggestion
                kind == SyntaxKind.QUALIFIED_NAME_REFERENCE ||
```

File: openapi-bal-service/src/main/java/io/ballerina/openapi/converter/service/OpenAPIRequestBodyMapper.java [hunk 2]
```diff
@@ -152,62 +160,176 @@ public void handlePayloadAnnotation(RequiredParameterNode payloadNode, Map<Strin
...
+                    handleReferencePayload(typeSymbol, mimeType, recordName, schema, bodyParameter);
+                    break;
+                case ARRAY_TYPE_DESC:
+                    ArrayTypeDescriptorNode arrayNode = (ArrayTypeDescriptorNode) payloadNode;
+                    mediaTypeString = getMediaTypeForSyntaxKind(arrayNode.memberTypeDesc());
+                    handleArrayTypePayload(schema, arrayNode, mediaTypeString, bodyParameter);
+                    break;
+                default:
+                    //Warning message for unsupported request payload type in Ballerina resource.
+                    DiagnosticMessages errorMessage = DiagnosticMessages.OAS_CONVERTOR_116;
```
Comment (dilanSachi) [inline openapi-bal-service/src/main/java/io/ballerina/openapi/converter/service/OpenAPIRequestBodyMapper.java:245]: We dont need an extra variable here right?

File: openapi-bal-service/src/main/java/io/ballerina/openapi/converter/service/OpenAPIRequestBodyMapper.java [hunk 3]
```diff
@@ -152,62 +160,176 @@ public void handlePayloadAnnotation(RequiredParameterNode payloadNode, Map<Strin
...
+                    IncompatibleResourceDiagnostic error = new IncompatibleResourceDiagnostic(errorMessage,
+                            payloadNode.location(), String.valueOf(payloadNode.kind()));
+                    diagnostics.add(error);
+                    break;
+            }
+        } else {
+            //Warning message for unsupported request payload type in Ballerina resource.
+            DiagnosticMessages errorMessage = DiagnosticMessages.OAS_CONVERTOR_116;
```
Comment (dilanSachi) [inline openapi-bal-service/src/main/java/io/ballerina/openapi/converter/service/OpenAPIRequestBodyMapper.java:253]: here also

File: openapi-bal-service/src/main/java/io/ballerina/openapi/converter/service/OpenAPIRequestBodyMapper.java [hunk 4]
```diff
@@ -234,19 +357,36 @@ private void handleArrayTypePayload(Map<String, Schema> schema, ArrayTypeDescrip
...
+                content.addMediaType(mimeType, media);
             }
+            content.addMediaType(mimeType, media);
+            requestBody.setContent(content);
+        } else {
+            requestBody.setContent(new Content().addMediaType(mimeType, media));
+        }
+        //  Adding conditional check for http delete operation as it cannot have body
+        //  parameter.
+        if (!operationAdaptor.getHttpOperation().equalsIgnoreCase("delete")) {
```
Comment (dilanSachi) [inline openapi-bal-service/src/main/java/io/ballerina/openapi/converter/service/OpenAPIRequestBodyMapper.java:388]: Do we have any constants already defined for "delete"? If so we can use it

File: openapi-bal-service/src/main/java/io/ballerina/openapi/converter/service/OpenAPIRequestBodyMapper.java (hunk 2 above)
Comment (lnash94) [reply openapi-bal-service/src/main/java/io/ballerina/openapi/converter/service/OpenAPIRequestBodyMapper.java:245]: Yeah, will assign it to directly

File: openapi-cli/src/test/resources/ballerina-to-openapi/request_body/union_type.bal [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+public type ABC record {
+    int id;
+    string name;
+};
+
+service /payloadV on new http:Listener(0) {
+    resource function post path(ABC? payload) {
+    }
+    resource function post path2(ABC|xml payload) {
```
Comment (TharmiganK) [inline openapi-cli/src/test/resources/ballerina-to-openapi/request_body/union_type.bal:11]: How about this scenario? Do we support this?
```bal
type UnionType ABC|xml;

...

resource function post path(UnionType payload) {
```

File: openapi-bal-service/src/main/java/io/ballerina/openapi/converter/service/OpenAPIRequestBodyMapper.java (hunk 4 above)
Comment (lnash94) [reply openapi-bal-service/src/main/java/io/ballerina/openapi/converter/service/OpenAPIRequestBodyMapper.java:387]: Fixed here: https://github.com/ballerina-platform/openapi-tools/pull/1544/commits/213d3e3838ce199712f5d0a24d321617db6d99c2

File: openapi-cli/src/test/resources/ballerina-to-openapi/request_body/union_type.bal (hunk 5 above)
Comment (lnash94) [reply openapi-cli/src/test/resources/ballerina-to-openapi/request_body/union_type.bal:15]: Yeah, we need to address this scenario, its logic may complicated and we wish to address this in the separate PR. For tracking purposes, I added the tests there  https://github.com/ballerina-platform/openapi-tools/pull/1544/commits/127fc070dec352e23a9ac62a832045e41bc8ff2a.

**Review by TharmiganK — APPROVED** [review]
LGTM

File: openapi-cli/src/test/resources/ballerina-to-openapi/request_body/union_type.bal [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    resource function post path6(@http:Payload int|string payload) {
+    }
+    resource function post path7(@http:Payload map<json>|int[] payload) {
+    }
+    resource function post path8(@http:Payload map<int>|string payload) {
+    }
+    resource function post path9(@http:Payload map<int>|map<string> payload) {
+    }
+
+    // //negative - skip
```
Comment (ayeshLK) [inline openapi-cli/src/test/resources/ballerina-to-openapi/request_body/union_type.bal:33]: Better to remove the commented code.

Comment (lnash94) [reply openapi-cli/src/test/resources/ballerina-to-openapi/request_body/union_type.bal:33]: This is added with the purpose of adding negative tests in the future.

**Review by ayeshLK — APPROVED** [review]

File: openapi-bal-service/src/main/java/io/ballerina/openapi/converter/service/OpenAPIRequestBodyMapper.java [hunk 7]
```diff
@@ -152,62 +161,174 @@ public void handlePayloadAnnotation(RequiredParameterNode payloadNode, Map<Strin
...
+                            DiagnosticMessages.OAS_CONVERTOR_116, payloadNode.location(),
+                            String.valueOf(payloadNode.kind()));
+                    diagnostics.add(error);
+                    break;
+            }
+        } else {
+            //Warning message for unsupported request payload type in Ballerina resource.
+            IncompatibleResourceDiagnostic error = new IncompatibleResourceDiagnostic(
+                    DiagnosticMessages.OAS_CONVERTOR_116, payloadNode.location(), payloadNode.toSourceCode().trim());
+            diagnostics.add(error);
```
Comment (NipunaRanasinghe) [inline openapi-bal-service/src/main/java/io/ballerina/openapi/converter/service/OpenAPIRequestBodyMapper.java:255]: Here we are using the same error, but with two different properties of the payload (payload kind at L247 and, payload node itself at L254). Is that intentional?

File: openapi-bal-service/src/main/java/io/ballerina/openapi/converter/service/OpenAPIRequestBodyMapper.java [hunk 8]
```diff
@@ -290,23 +429,52 @@ private void handleReferencePayload(TypeSymbol typeSymbol, String recordName,
...
         //  parameter.
-        if (!operationAdaptor.getHttpOperation().equalsIgnoreCase("delete")) {
+        if (!operationAdaptor.getHttpOperation().equalsIgnoreCase(DELETE)) {
             operationAdaptor.getOperation().setRequestBody(bodyParameter);
         }
     }
 
-    private void addConsumes(OperationAdaptor operationAdaptor, RequestBody bodyParameter, String applicationType) {
-        String type = applicationType.split("/")[1];
+    private void addConsumes(OperationAdaptor operationAdaptor, RequestBody bodyParameter, String applicationType,
```
Comment (NipunaRanasinghe) [inline openapi-bal-service/src/main/java/io/ballerina/openapi/converter/service/OpenAPIRequestBodyMapper.java:454]: Whats the purpose of this util?

File: openapi-bal-service/src/main/java/io/ballerina/openapi/converter/service/OpenAPIRequestBodyMapper.java [hunk 9]
```diff
@@ -234,19 +356,36 @@ private void handleArrayTypePayload(Map<String, Schema> schema, ArrayTypeDescrip
...
             arraySchema.setItems(itemSchema.$ref(ConverterCommonUtils.unescapeIdentifier(
                     referenceNode.name().text().trim())));
-            io.swagger.v3.oas.models.media.MediaType media = new io.swagger.v3.oas.models.media.MediaType();
             media.setSchema(arraySchema);
-            if (requestBody.getContent() != null) {
-                Content content = requestBody.getContent();
-                content.addMediaType(mimeType, media);
+        } else if (typeDescriptorNode.kind() == SyntaxKind.BYTE_TYPE_DESC) {
+            StringSchema byteSchema = new StringSchema();
+            byteSchema.setFormat("byte");
```
Comment (NipunaRanasinghe) [inline openapi-bal-service/src/main/java/io/ballerina/openapi/converter/service/OpenAPIRequestBodyMapper.java:362]: Shall we reuse our predefined constant for "byte" in here?

**Pull request merged**